draft-ietf-quic-transport.md

```diff
@@ -1069,6 +1069,15 @@ to cease using the connection IDs when requested can result in connection
 failures, as the issuing endpoint might be unable to continue using the
 connection IDs with the active connection.
 
+An endpoint SHOULD limit the number of outstanding RETIRE_CONNECTION_ID frames
+to bound the necessary state. In order to allow a peer to retire all previously
+issued connection IDs, the limit on the number of outstanding
+RETIRE_CONNECTION_IDs SHOULD be at least the active_connection_id_limit. An
```

Comment:
How about something like: _In order to allow a peer to retire all previously issued connection IDs while an endpoint waits for some RETIRE_CONNECTION_ID frames to be acknowledged, the limit on the number of outstanding RETIRE_CONNECTION_IDs SHOULD be at least the active_connection_id_limit plus the number of retirements that the endpoint might initiate at one time._

Reply:
I don't think that is sufficient or necessary.

1. sufficiency: an adversary can send a long sequence of NEW_CONNECTION_ID frames with seq: N and rpt: N-1.  This is both entirely legitimate and completely outside the spirit of the protocol.

2. necessity: if you track min(unacknowledged RCID) and max(RPT) then you can track as few RCID frames as you like.  It might be nice if people used active_connection_id_limit*2, but that number only optimizes for a complete turnover of connection IDs every round trip.  It is an otherwise arbitrary number.  I don't think that we need to optimize for that case.

The only problem arises with this approach is when that gap widens more than you would like.  At which point you can throw down a CONNECTION_CLOSE and walk away.

Reply:
It occurred to me that we might be talking about two entirely different ideas.

1) The "Maximum number of outstanding RCIDs" is a momentary restriction. I might send only 10 at a time, or whatever, but eventually I will send all of the RCIDs that the protocol required before any of these PRs.

2) I need only send N RCIDs at a time, and can assume all other (earlier) sequence numbers have been de facto retired.

If my limit is 10 and I get 100 NCIDs that increment RPT by 1 in rapid succession, option (1) would have us send all 100 RCIDs over 10 RTTs, assuming no loss. option (2) would have us send 10 and be done with it. Which are we describing here?

Reply:
I definitely mean option 1.  Which leads to eventual resolution.

Note that if we also take Ian's #3550, then none of these RCID frames are necessary.  You only use RCID for discretionary use (where you use a connection ID and find you no longer need it).  That's why I think that's a nice addition.

Reply:
I agree that we do not need to care about sufficiency.

> if you track min(unacknowledged RCID) and max(RPT) then you can track as few RCID frames as you like. It might be nice if people used active_connection_id_limit*2, but that number only optimizes for a complete turnover of connection IDs every round trip.

This observation is what I disagree with. The intent of the change I proposed right above is to allow an endpoint to retire all the CID at once, without being concerned about the peer losing track of some sequence numbers carried by CID frames already inflight, or the peer considering the endpoint to be doing something malicious.

Assuming that an endpoint that retires CIDs track each of the RCID frame it sends, the safe number is _active_connection_id_limit_ (the number of CIDs that might be retired at once by RPT) plus _the number of retirements that that endpoint might initiate at once_.

To give a specific example, consider the following case.
* A client, with max_connection_id_limit = 2, is using CID<sub>0</sub> for the main path, and is probing a new path using CID<sub>1</sub>.
* Server sends NCID(CID<sub>2</sub>, RPT=2), NCID(CID<sub>3</sub>, RPT=2).
* Client receives these NCID frames, and assigns CID<sub>2</sub> to the main path, CID<sub>3</sub> to the path being probed. The client sends RCID(CID<sub>0</sub>), RCID(CID<sub>1</sub>).
* Before receiving ACK for the two RCID frames, the client decides to abandon the path being probed. It needs to send RCID(CID<sub>2</sub>).

At this point, the client is required to track the flight of RCID frames carrying *three* CIDs. And it has to make sure that RCID(CID<sub>2</sub>) reaches the peer, otherwise it would not receive a CID that substitutes CID<sub>2</sub>.

The recommendation in this PR now (i.e. "the limit on the number of outstanding RETIRE_CONNECTION_IDs SHOULD be at least the active_connection_id_limit") fails to cover this case.

Reply:
@martinthomson 
> Note that if we also take Ian's #3550, then none of these RCID frames are necessary. You only use RCID for discretionary use (where you use a connection ID and find you no longer need it). That's why I think that's a nice addition.

I think I agree with this. Status quo is just confusing, as it requires endpoints to send RCID frames for the retired CIDs they have received, but does not require the frames to be sent for the CIDs that they receive afterwards.

This means that from the viewpoint of the endpoint that requests retirement, it cannot assume that it would see RCID frames for the CIDs that it has requested retirement, but in most cases it would see them. This could turn into a trap in some implementations.

I agree that we should adopt #3550, so that we would never assume to see RCID for the CIDs that we request retirement.

Reply:
I see what you mean by the CID<sub>2</sub> example, but my counter would be that you wouldn't retire that connection ID because you were still busy working through the previous set of retirements.  (This assumes that you don't or can't start to retire CID<sub>2</sub> before being told to retire the others.)

The problem with discretionary retirement is that it won't be contiguous in the same way that Retire Prior To has to be, so you potentially have a patchwork of connection IDs that are variously valid, in use, unused, or retired.  You can use run-length encoding to reduce state somewhat, but there is no easy way to compress this state commitment.

But at least you can avoid the costs of retiring stuff that you are forced to retire.  That's all I was worried about.  Costs of discretionary activities are something people just have to factor in.

Reply:
> I see what you mean by the CID<sub>2</sub> example, but my counter would be that you wouldn't retire that connection ID because you were still busy working through the previous set of retirements. (This assumes that you don't or can't start to retire CID<sub>2</sub> before being told to retire the others.)

Yeah I think my view is that it is hard to implement such behavior without having additional state, because the peer can request an endpoint to retire CIDs. You can delay the start to retire CID<sub>2</sub>. But to do that, you need to have a delayed slot (i.e. state).

> Costs of discretionary activities are something people just have to factor in.

I agree. My complaint is that the recommendation of "at least active_connection_id_limit" is incorrect.

As we seem to be converging on adopting #3550, I've posted an example that does not involve the use of RPT in https://github.com/quicwg/base-drafts/issues/3509#issuecomment-606408197. I posted it on the issue as I think it is a part of the problem statement rather than the solution.

Reply:
Alright, I implemented this PR this morning and it's quite lightweight. Moreover, it's fully interoperable with draft-27.

#3550 is a much more drastic change in behavior that will break interoperability. draft-27 NCID senders will wait for RCIDs that will never come. If we're going to make a breaking change I'd just as soon we go to #3553 and have a comprehensive solution that is less likely to end up with corner cases.

Reply:
Q: At what threshold did you close the connection?  2 * active_connection_id_limit?  Or did you not close the connection and instead do something else?

Reply:
So we're limiting in-flight to 2*(the limit), which for us is currently 4. I set a very large threshold to tear down the connection -- currently, if highest_rpt is 100 more than least_unacked. That makes it exceedingly unlikely weird ack loss cases are going to bring me down, but I can tear down the connection if it gets ridiculous. There's really no difference in state between it being 50, 100, or 1000 -- it just means the client will get application data a little slower.

Reply:
@martinduke Thank you for sharing your experience.

Could you please clarify why you have two limits? Are you limiting the number of inflight RCID frames to 4, even though you retain state of up to 100 CIDs to be retired?

I'm curious, because that's not something I would expect as an outcome of this PR. My read of the text proposed in the PR has been that we assume that a RCID frame becomes "outstanding" when an endpoint decides to (or being instructed to) retire a CID.

IIUC, what we need to do is limit the amount of state that the endpoint retains. Number of RCID frames sent at one time does not really matter, as it is bound by the CWND size.

Reply:
@kazuho. I initially read "outstanding" as meaning "intended to retire", just like you, but further up in the thread @martinthomson said that's not what he meant. And anyway, I soon saw that that interpretation is equally easy to implement and more compliant with the existing spec.

First, assume the endpoint never retires a CID voluntarily. I just have to store the active CIDs. The retired CIDs require only two state variables:
- highest retire prior to (highest_rpt)
- lowest RCID seqnum unacked (lowest_rua)

Each packet records the highest seqnum retired because it always retires everything below that, as you'll see below.

`process_ncid():
    if (retire_prior_to <= highest_rpt)
        quit;
    if (retire_prior_to > lowest_rua + 100)
       abort;
    if (highest_rpt == lowest_rua)
        highest_rpt = retire_prior_to
        send_rcid_pkt()
    else
        highest_rpt = retire_prior_to

send_rcid_pkt():
    limit = MIN(highest_rpt, lowest_rua + 2 * active_cid_limit)
    for i = lowest_rua : limit
        send_rcid(i)
    pkt.max_seq_num = limit

process_ack(pkt):
    lowest_rua = MAX(lowest_rua, pkt.max_seq_num + 1) 
    if (lowest_rua < highest_rpt) 
        send_rcid_pkt()`

If I have a backlog of CIDs to retire, I will clear 2 * active_cid_limit of them each RTT assuming no losses. This number is essentially arbitrary but should definitely fit in a single packet. 100 is an arbitrarily large number just in case the client is being malicious. Again, you don't want it to be so small that unlucky packet losses cause a connection error, but beyond that any number is fine.

If we are retiring CIDs voluntarily, we'll have to keep some extra state. We can put them in a special data structure that send_rcid_pkt also checks, but will definitely be finite as it is controlled locally. If I retire them in sequence, nothing stops me from increasing highest_rpt even though I never received it in an NCID frame, and therefore using the mechanism above.

***

The nice thing about this design is that we are fully compliant with the draft-27 language in every way, except that the RCID frames are not always sent *immediately*. #3553 is a cleaner design but this one has essentially no impact on the peer at all if it's implementing -27.

Reply:
@martinduke Thank you for the clarification. I now understand the intent. I think that we should adjust the text in this PR.

To mitigate the attacks being discussed, what we need to do is cap the amount of state used for retaining the CIDs that "are intended to be retired." While an endpoint may restrict the number of RCID frames being sent at one time, that is unrelated to the attack, and therefore I would advocate for removing that requirement (as it is an unnecessary design change).

Speaking of your implementation, what matters is 100. 4 (2 * active_connection_id_limit) is an arbitrary choice unrelated to the attack; my argument is that therefore we do not need to specify that in spec.

Reply:
PS. To avoid the confusion being discussed in the comments right above, I think my preference goes to having something like:

_An adversarial peer can request an endpoint to retain state for an unlimited amount of connection IDs being retired by withholding acknowledgements for the RETIRE_CONNECTION_ID frames, at the same time sending NEW_CONNECTION_ID frames with updated values in the Retire Prior To fields, or probing and discarding paths repeatedly._

_As a mitigation, endpoints SHOULD limit the amount of state that they retain for retiring the connection IDs. In order to allow a well-behaving peer to request retirement of all the previously issued connection IDs while an endpoint waits for some RETIRE_CONNECTION_ID frames to be acknowledged,  an endpoint SHOULD be capable of tracking at least the active_connection_id_limit plus the number of retirements that the endpoint might initiate at one time. When the amount of state required for tracking the connection IDs being retired exceeds that limit, an endpoint MAY close the connection with a CONNECTION_ID_LIMIT_ERROR._

Reply:
@kazuho I believe my psuedocode indicates that the current version of the PR doesn't add more state for various values of "maximum RCIDs inflight" as long as it's small enough that you can fit it all in one QUIC packet. If the RCIDs are spread over multiple packets inflight then keeping track of RCIDs requires a scoreboard.

The exact threshold to throw an error is indeed not needed in the spec. However @ianswett, rightfully, requested a MAY around the error to set expectations for debugging.

I do think we should change the word "outstanding" to "in flight" just to make it a bit clearer.

Reply:
in flight or unacknowledged would both be fine with me.  One negative of "in flight" is that if you're congestion blocked, you can't yet put the RCID in flight.

Reply:
unacknowledged is better

Reply:
"unacknowledged" would also include long-lost packets, technically. I agree with the congestion control observation, but I'm not sure how it impacts the behavior. I'll _always_ withhold RCIDs that are cwnd-blocked; it's a restriction on RCIDs in flight independent of this one.

Reply:
@martinduke I agree that when restricted by CWND, we cannot send a frame. Therefore my view is that it is difficult or confusing to call a frame that is not yet being set as "inflight" or "outstanding." Both words are wrong IMO.

That the reason why my text above talks about "state retained for retiring connection IDs."

> The exact threshold to throw an error is indeed not needed in the spec. However @ianswett, rightfully, requested a MAY around the error to set expectations for debugging.

While I am not against making such a suggestion using MAY in text, if we are to make a suggestion, I think we'd need to be very cautious so that the limit would not be too small.

Reply:
Perhaps I'm missing something obvious. The endpoint can say "I will only have 10 RCIDs in flight at a time". So if the peer makes me send 20, it'll take me at least 2 RTTs to resolve. Now obviously it can take longer if I don't have enough cwnd, but that's orthogonal to this and no change to the spec. Yes?

I'm not sure why an endpoint would choose a threshold that would result in spurious aborts, but if there's some limit you would suggest I'm happy to add it.

Reply:
> The endpoint can say "I will only have 10 RCIDs in flight at a time".

Yes. That's possible, but having that limitation does not mitigate the attack.

> So if the peer makes me send 20, it'll take me at least 2 RTTs to resolve. Now obviously it can take longer if I don't have enough cwnd, but that's orthogonal to this and no change to the spec.

The attack is that peer might make you send more than 20 or more. And we need to either provide a guidance on when an endpoint can start leaking retirements or close the connection.

However, the only SHOULD being added in this PR is something unrelated to the attack. And the location of the text gives the false impression that the limit in that sentence is the threshold for mitigating the attack.